src/Inciter/AMR/node_store.h

```diff
@@ -0,0 +1,245 @@
+                    return static_cast<size_t>(already_exists);
+                }
+
+            }
+
+            /**
+             * @brief Function to add a new node
+             *
+             * @param coord_tuple The coordinate data to add for the node
+             *
+             * @return Id of node added
+             */
+            size_t add(coordinate_t coord_tuple)
+            {
+                return add( coord_tuple[0], coord_tuple[1], coord_tuple[2]);
+            }
+
+            /**
+             * @brief Function to add a new point/coordinates
+             *
+             * @param x x val
+             * @param y y val
+             * @param z z val
+             *
+             * @return id of coordinate added
+             */
+            size_t add_coordinates(real_t x, real_t y, real_t z) {
+                add_x(x);
+                add_y(y);
+                add_z(z);
+                return size()-1; // -1 because of the 0 index
+            }
+
+            /**
+             * @brief Function to add a new node
+             *
+             * @param x x val of node
+             * @param y y val of node
+             * @param z z val of node
+             *
+             * @return id of node added
+             */
+            size_t add_node(real_t x, real_t y, real_t z) {
+
+                // Need to: Add to {x,y,z} Add any connectivity?
+
+                // Check if the node already exists
+                int already_exists = check_node_exists(x,y,z);
+
+                if (already_exists == -1) {
+                    size_t return_node_id = add_coordinates(x,y,z);
+                    (*m_graphsize)++; // TODO: how to deal with this?
+                    trace_out << "--> Made " << return_node_id << std::endl;
+                    return return_node_id;
+                }
+                else {
+                    trace_out << "--> Reusing " << already_exists << std::endl;
+                    return static_cast<size_t>(already_exists);
+                }
+
+            }
+
+            // TODO: Remove all calls to this as it's fairly expensive...
+            // TODO: Find a more cost effective way to implement this
+                // Most likely change data structure for a faster search
+                // This is also going to be a potential problem in async parallel
+            /**
+             * @brief Helper function to check if a node already exists at
+             * coords {x,y,z} to avoid it being duplicated
+             *
+             * @param x_in X coord to check
+             * @param y_in Y coord to check
+             * @param z_in Z coord to check
+             *
+             * @return The id of the node if it exists, -1 if it doesn't.
+             */
+            int check_node_exists(real_t x_in, real_t y_in, real_t z_in)
+            {
+                const real_t eps = 1e-7;
+                for (size_t i = 0; i < size(); i++)
```

Comment:
Yep, as the comments above indicate, this is a linear search and we probably want steer clear of it.